### PR TITLE
[Diagnostics & Tests] Add diagnostics for detecting invalid conversions from FP to Ints.

### PR DESCRIPTION
**Pull request by ravikandhadai**
Add test cases for checking the correctness of the diagnostics.

### CODE REVIEW

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test

**Review by eeckstein — CHANGES_REQUESTED** [review]
basically lgtm.

Can you please also add a SIL test for the constant propagation pass? E.g. add some test cases in test/SILOptimizer/constant_propagation.sil

**Comment by shajrawi** [comment]
Looks good - but as Erik said: please add some tests in `test/SILOptimizer/constant_propagation.sil`

File: test/SILOptimizer/diagnostic_constant_propagation_floats.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// on floating-point operations
+
+import StdlibUnittest
+
+func testFPToIntConversion() {
+  _blackHole(Int8(-1.28E2))
+  _blackHole(Int8(-1.27E2))
+  _blackHole(Int8(-0))
+  _blackHole(Int8(3.33333))
+  _blackHole(Int8(-2E2)) // expected-error {{floating-point literal '-200' overflows 8-bit signed integer type}}
```
Comment (stephentyrone) [inline test/SILOptimizer/diagnostic_constant_propagation_floats.swift]: These error messages could be a lot better if they printed the actual literal and type that occur in the source. In fact, the user doesn't care that it was a floating-point literal at all, so ideally we should just say something like "This conversion will always fail: `-2E2` overflows `Int8`."

**Review by stephentyrone — CHANGES_REQUESTED** [review]

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

File: test/SILOptimizer/diagnostic_constant_propagation_floats.swift (hunk 1 above)
Comment (ravikandhadai) [reply test/SILOptimizer/diagnostic_constant_propagation_floats.swift]: @stephentyrone Thanks for the comment.  I have updated the diagnostic messages to display the literal as in the source file, and use the source type, whenever possible. Let me know if the messages look okay now.

**Comment by ravikandhadai** [comment]
@eeckstein Thanks for the comments. I have added some SIL tests now (test/SILOptimizer/diagnostic_constant_propagation_floats.sil) and improved error reporting.

**Comment by stephentyrone** [comment]
Much improved, thanks very much!

**Review by stephentyrone — APPROVED** [review]

**Comment by ravikandhadai** [comment]
@eeckstein I see. Thanks for the clarification.  I will add them.
Btw, I guess the SIL tests I have currently are also useful to test diagnostics when AST information is not available. (The AST information is used to recover the source literal and type).  So, I would like to keep this too, if it is okay.

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test

**Comment by ravikandhadai** [comment]
@eeckstein Thanks for the comments. I have added some SIL tests that checks the correctness of the folded program obtained after constant propagation. The tests are available in the new file: test/SILOptimizer/constant_propagation_floats.sil

**Review by eeckstein — APPROVED** [review]
perfect, thanks!

**Comment by ravikandhadai** [comment]
@swift-ci Please test merge

**Comment by ravikandhadai** [comment]
@swift-ci  Please test and merge

**Comment by shahmishal** [comment]
@swift-ci Please test and merge

**Comment by shahmishal** [comment]
@swift-ci Please test and merge

**Pull request merged**